### PR TITLE
RTL support based on ICU library.

### PR DESCRIPTION
**Pull request by mrezai**
Hi
If the ICU [license](http://source.icu-project.org/repos/icu/icu/trunk/license.html) is acceptable, I think this PR can be considered as a reference implementation to find out the best approach to implement RTL support.
This feature is disable by default. Use rtl=yes in command line to enable it.
Text rendering and alignment implemented only for Button and Label but other controls(TextEdit, LineEdit, RichTextLabel) only have text rendering and not alignment.
I didn't removed unused files from icu source because maybe some of them can be use later.
Compilation tested for linux, windows(mingw) and android.

How to test:
- Compile editor and export templates with rtl=yes option.
- Add a RTL keyboard(language) to your device.
- Import a suitable font.Enable unicode when importing.
- Add font to Settings -> Editor Settings : Global/Font, Text Editor/Font.
- Restart Editor.
- Add some of controls mentioned above that support RTL.
- Change controls font to imported one.
- Finally, Test it!

This is based on mohaalak and masoudbh3 previous works. Thank you.

### CODE REVIEW

**Comment by akien-mga** [comment]
Related to #982 and #2297. Ping @masoudbh3 and @mohaalak for potential review :)

**Comment by akien-mga** [comment]
License-wise it looks very similar to the MIT (the only difference is that it asks that the license is also included in "supporting documentation"), so I guess it's fine.

**Comment by mohaalak** [comment]
this pull request has conflicts I cannot check it can you correct this pull request , I hope in version 2.0 we have RTL support .

**Comment by mrezai** [comment]
PR rebased. There is a problem in arabic harakat marks  shaping, maybe I used library incorrectly but I think for a general solution we must replace shaping part with harfbuzz to support other languages too.

**Comment by mohaalak** [comment]
everything looking great, I think it should be included in version 2

**Comment by mohaalak** [comment]
just need a little fix for conflict.  @mrezai thanks

**Comment by reduz** [comment]
It's too late for 2.0, also I would prefer we just strip the code for rtl
instead of linking the whole library

just need a little fix for conflict. @mrezai <https://github.com/mrezai>
thanks

—
Reply to this email directly or view it on GitHub
<https://github.com/godotengine/godot/pull/2886#issuecomment-183907776>.

**Comment by mrezai** [comment]
Rebased. @mohaalak, I mentioned the harakat mark shaping problem before.
@reduz, yeah this isn't good enough to merge. I'll write my opinion about i18n implementation later in #3081 .

File: core/translation.cpp [hunk 1]
```diff
@@ -260,6 +262,8 @@ static const char* locale_names[]={
 "Spanish (Venezuela)",
 "Estonian",
 "Estonian (Estonia)",
+"Farsi",
+"Farsi (IRAN)",
```
Comment (bojidar-bg) [inline core/translation.cpp]: I guess you shouldn't write it in all-caps BTW.

**Comment by akien-mga** [comment]
@mrezai Are you still up for working on this? 2.1 has i18n support now, so having a good RTL implementation would be really nice :D

**Comment by abdlquadri** [comment]
@mrezai Any help needed on this?

**Comment by reduz** [comment]
I found that Unicode has a simple reference implementation:

http://www.unicode.org/Public/PROGRAMS/BidiReferenceCpp/v25g/

think anyone could adapt it?

**Comment by abdlquadri** [comment]
@reduz I read the code overnight. Also studied @mrezai's implementation with ICU. But I still not clear where in the godot code base this will impact. basically where will I be needing to make changes..

**Comment by akien-mga** [comment]
Closing as this PR can't be merged as is (conflicts, unresolved issues), but it's a great reference for a new attempt at implementing RTL support. Let's continue the discussion in #982.

**Pull request closed without merging**